docs/features/async-streams.md

```diff
@@ -164,3 +170,135 @@ Similarly, the kick-off method is much like those of regular iterator methods:
     return result;
 }
 ```
+
+#### Disposal
+
+Iterator and async-iterator methods need disposal because their execution steps are controlled by the caller, which could choose to dispose the enumerator before getting all of its elements.
+For example, `foreach (...) { if (...) break; }`.
+In contrast, async methods continue running autonomously until they are done. They are never left suspended in the middle of execution from the caller's perspective, so they don't need to be disposed.
+
+In summary, disposal of an async-iterator works based on four design elements:
+- `yield return` (jumps to finally when resuming in dispose mode)
+- `yield break` (enters dispose mode and jumpps to finally)
+- `finally` (after a `finally` we jump to the next one)
+- `DisposeAsync` (enters dispose mode and resumes execution)
+
+The caller of an async-iterator method should only call `DisposeAsync()` when the method completed or was suspended by a `yield return`.
+`DisposeAsync` sets a flag on the state machine ("dispose mode") and (if the method wasn't completed) resumes the execution from the current state.
+The state machine can resume execution from a given state (even those located within a `try`).
+When the execution is resumed in dispose mode, it jump straight to the relevant `finally`.
+`finally` blocks may involve pauses and resumes, but only for `await` expressions. As a result of the restrictions imposed on `yield return` (described above), dispose mode never runs into that statement.
+Once a `finally` block completes, the execution in dispose mode jumps to the next relevant `finally`, or the end of the method once we reach the top-level.
+
+Reaching a `yield break` also sets the dispose mode flag and jumps to the next relevant `finally` (or end of the method). By the time we return control to the caller (completing the promise as `false` by reaching the end of the method) all disposal was completed, and the state machine is left in finished state. So `DisposeAsync()` has no work left to do.
+
+Looking at disposal from the perspective of a given `finally` block, the code in that block can get executed:
+- by normal execution (ie. after the code in the `try` block),
+- by raising an exception inside the `try` block (which will execute the necessary `finally` blocks and terminate the method in Finished state),
+- by calling `DisposeAsync()` (which resumes execution in dispose mode and jumps to the relevant finally),
+- following a `yield break` (which enters dispose mode and jumps to the relevant finally),
+- in dispose mode, following a nested `finally`.
+
+A `yield return` is lowered as:
+```C#
+_current = expression;
+_state = <next_state>;
+goto <exprReturnTruelabel>; // which does _valueOrEndPromise.SetResult(true); return;
+
+// resuming from state=<next_state> will dispatch execution to this label
+<next_state_label>: ;
+this.state = cachedState = NotStartedStateMachine;
+if (disposeMode) /* jump to current finally or exit */
```

Comment:
>/* [](start = 17, length = 2)

Minor point: perhaps `//`

Reply:
I've used the following convention: `/* ... */` as a placeholder for some omitted code, but use `// ...` to comment on code that is there.

---
In reply to: [240709778](https://github.com/dotnet/roslyn/pull/31527#discussion_r240709778) [](ancestors = 240709778)